rules/do-you-optimize-tinacms-project-for-clarity-performance-and-reliable-builds/rule.md

```diff
@@ -0,0 +1,109 @@
+
+* **Caching Data at a Top-Level and Accessing It When Necessary:**
+
+  * If passing props is not feasible (e.g., when a component depends on Next.js router information), you should make a general top-level request, cache the data, and then access it directly from the cache within the component.
+  * This approach ensures efficient data retrieval and reduces the server load at build time.
+
+## 2. Improving Runtime Performance
+
+Optimizing runtime performance is key to delivering a fast and responsive user experience.
+
+### Bad Practices:
+
+* **Using Client-Side Requests Instead of Cached Values from Server-Side Components:**
+
+  * This approach bypasses pre-fetched, cached data, increasing server load and slowing down the application.
+  * It negates the benefits of static site generation, where data should be ready and waiting.
+
+### Common Practices:
+
+* **Utilizing Next.js caching via fetch:** 
+
+  * Next.js’s fetch API is designed to cache by default during static generation and server-side rendering.
+
+### Good Practices:
+
+* **Utilizing Caching at the Highest Level of the Application:**
+
+  * By caching data at the top level, you ensure that it is readily available throughout your application, reducing redundant data fetching and improving overall performance.
+  * This approach allows for more efficient data management and quicker response times.
+* **Using Dedicated Caching Packages:**
+
+  * [**Redis**:](https://redis.io/fr/solutions/cas-dutilisation/cache/) An in-memory data structure store, Redis is fast and versatile, making it ideal for caching data in memory. 
+  * **[Node-cache](https://www.npmjs.com/package/node-cache)**: A lightweight, in-memory caching solution for Node.js applications.
+
+### Gold Plating:
+
+* **Implementing a Singleton Service for Data Requests:**
+
+  * A singleton service that encapsulates general requests and provides data via cache ensures consistent data retrieval and minimizes repetitive API calls, thereby enhancing performance.
+
+## 3. Preventing Build Crashes
+
+To ensure smooth and reliable builds, it’s important to follow best practices that prevent excessive server load and manage data efficiently.
+
+### Best Practices:
+
+* **Avoid Isolating Tina Client Calls in Multiple Subcomponents:**
+
+  * Centralize data fetching to avoid overwhelming the build process with excessive API requests.
+* **Cache Large Datasets at a Top Level:**
+
+  * By caching large datasets at a top level, you reduce the number of data fetches during the build process, lowering the risk of overwhelming the system.
+* **Avoid Using Deeply Nested References:**
+
+  * Simplifying your data structure and avoiding deeply nested references can help ensure smoother and faster builds.
+* **Write custom Tina queries:**
+
+  * You can improve build generation and prevent build crash by creating your own Tina GraphQL queries.
```

Comment:
Not sure I agree with these. 

**Avoid Isolating Tina Client Calls in Multiple Subcomponents:** ✅ - The wording could be better but generally you call once and pass the result down. Could use prop drilling but potentially context could be better with deeply nested components. 

* Centralize data fetching to avoid overwhelming the build process with excessive API requests.* ❌ Not sure I fully under stand these? 
**Cache Large Datasets at a Top Level:** ❌ Not sure I fully under stand these? 

  * Simplifying your data structure and avoiding deeply nested references can help ensure smoother and faster builds.* ❌ This is more a programming best practice 
  
  **Write custom Tina queries:** 🚧 Can you give / write examples of these 🙏
  
  * You can improve build generation and prevent build crash by creating your own Tina GraphQL queries. - ❌ Not sure I fully understand

Reply:
The titles are definitely misleading if you don't have context of their problem

> Avoid Isolating Tina Client Calls in Multiple Subcomponents

they were making calls to Tina to retrieve the data for the website color scheme + header + footer for every page, I think that the caching mods will negate this issue

Reply:
Yes, Matt is right the text is laking some context.
I added some context/examples.

> Avoid Isolating Tina Client Calls in Multiple Subcomponents

As Matt mentioned, one of our cases is that on each page, we had a header and a footer that fetched data every time we navigated to a new page on the website. We improved performance by caching this data from a top-level call.

> I think that the caching mods will negate this issue

Should I delete this part ?

> Cache Large Datasets at a Top Level

In our context, on the NDC website, each time we clicked on a speaker to view information, we fetched data, resulting in numerous API calls during the build process, which led, in our case, to build failures. By caching all that data at the top level, we significantly reduced API calls during the build process (which also resolved our build crashes).

> Avoid Using Deeply Nested References:
❌ This is more a programming best practice

Should we get rid of it ?

> Write custom Tina queries:

Some of the auto-generated Tina queries were not optimized, such as nested objects with redundant data. For example, speakers that included a conference object, which in turn included the same speakers again. Creating custom queries can reduce the size of objects and improve performance.
Check for example that commit: https://github.com/SSWConsulting/NDC_Website/commit/2755d495e66c70b8c3cf2fe888b1828f9d4b41fb

Reply:
> Avoid Using Deeply Nested References:
> ❌ This is more a programming best practice
> 
> Should we get rid of it ?

Yea I reckon get rid of this part 

> I think that the caching mods will negate this issue

> Should I delete this part ?

✅ yep